akka-actor-typed/src/main/scala/akka/actor/typed/pubsub/PubSub.scala

```diff
@@ -0,0 +1,149 @@
+   *
+   * @tparam T the type of messages the topic accepts for publishing and subscribing
+   * @param ttl If the topic is idle this long, shut the actor down. All calls to `topic` for the same topic name should
+   *            use the same ttl.
+   */
+  def topic[T](name: String, ttl: FiniteDuration)(implicit classTag: ClassTag[T]): ActorRef[Topic.Command[T]] =
+    topicFor[T](name, Some(ttl))(classTag)
+
+  /**
+   * Java API: Spawn an actor with the given topic name or share an existing one if it is already running.
+   *
+   * Note that [[Topic]] actors manually started will not be part of this registry.
+   *
+   * @param messageClass the type of messages the topic accepts for publishing and subscribing
+   */
+  def topic[T](messageClass: Class[T], name: String): ActorRef[Topic.Command[T]] =
+    topicFor(name, None)(ClassTag(messageClass))
+
+  /**
+   * Scala API: Spawn an actor with the given topic name or share an existing one if it is already running.
+   *
+   * Note that [[Topic]] actors manually started will not be part of this registry.
+   *
+   * @tparam T the type of messages the topic accepts for publishing and subscribing
+   * @param ttl If the topic is idle this long, shut the actor down. All calls to `topic` for the same topic name should
+   *            use the same ttl.
+   */
+  def topic[T](messageClass: Class[T], name: String, ttl: java.time.Duration): ActorRef[Topic.Command[T]] =
+    topicFor(name, Some(ttl.toScala))(ClassTag(messageClass))
+
+  /**
+   * Scala API: return the current set of running topics
+   */
+  def currentTopics: Set[String] = registry.keySet().asScala.toSet
+
+  /**
+   * Java API: return the current set of running topics
+   */
+  def getCurrentTopics(): java.util.Set[String] = registry.keySet().stream().collect(Collectors.toSet[String])
+
+  private def topicFor[T](name: String, ttl: Option[FiniteDuration])(
+      classTag: ClassTag[T]): ActorRef[Topic.Command[T]] = {
+    val messageClass = classTag.runtimeClass
+    val (classInRegistry, ttlInRegistry, actorRefInRegistry) =
+      registry.computeIfAbsent(
+        name,
+        name => (messageClass.asInstanceOf[Class[Any]], ttl, spawnTopic(name, classTag, ttl)))
+    if (classInRegistry != messageClass)
+      throw new IllegalArgumentException(
+        s"Trying to start topic [$name] with command class [$messageClass], but it is was already started with a " +
+        s"different command class [$classInRegistry]. Use unique topic names for all your topics.")
+    if (ttlInRegistry != ttl)
+      log.warn(
+        "Asked for topic [{}] with TTL [{}] but already existing topic has a different TTL [{}]",
+        name,
+        ttl.map(_.pretty).getOrElse("none"),
+        ttlInRegistry.map(_.pretty).getOrElse("none"))
+    actorRefInRegistry.narrow
+  }
+
+  private def spawnTopic[T](name: String, classTag: ClassTag[T], ttl: Option[FiniteDuration]): ActorRef[Any] = {
```

Comment:
maybe we can make it less untyped? here is a suggestion that compiles a0aa9a9843

Reply:
Included